refactor: Extract state and validation logic

pereng11 opened the pull request:
### 1️⃣ PR 타입(하나 이상의 PR 타입을 선택해주세요)
- [x] 기능 추가
- [ ] 기능 삭제
- [x] 버그 수정
- [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트
<br>

### 3️⃣ 변경 사항
- value / isDirty / isValid를 값으로 하는 상태를 useInput customHook으로 분리하였습니다
- validation과 관련한 로직을 utils/validate.ts로 분리하였습니다
- 로직을 추출함에 따라 props가 일부 변경되었습니다
  - value / isValid prop 삭제
  - initialValue prop 추가
- size prop의 값이 sm | md인 경우 label 태그를 숨김처리하는 방식을 변경하였습니다
- 에러 메시지 공간이 확보되지 않는 버그를 수정하였습니다

pereng11 committed:
refactor: Extract state and validation logic

- value / isDirty / isValid를 값으로 하는 상태를 useInput customHook으로 분리하였습니다
- validation과 관련한 로직을 utils/validate.ts로 분리하였습니다
- 로직을 추출함에 따라 props가 일부 변경되었습니다
  - value / isValid prop 삭제
  - initialValue prop 추가
- size prop의 값이 sm | md인 경우 label 태그를 숨김처리하는 방식을 변경하였습니다
- 에러 메시지 공간이 확보되지 않는 버그를 수정하였습니다

pereng11 committed:
Merge branch 'develop' into Input-Jaeyeong

CHAERYEON-LEE commented on src/components/common/Input/Input.module.css:
```diff
@@ -44,17 +44,11 @@
   height: 39px;
   width: 250px;
-  padding-left: 20px;
+  /* padding-left: 20px; */
```
굳굳 이거 그냥 지워도 될것같슴다

CHAERYEON-LEE commented on src/components/common/Input/Input.module.css:
```diff
@@ -65,16 +59,10 @@
   height: 41px;
   width: 343px;
-  padding-left: 22px;
+  /* padding-left: 22px; */
```
이것도 지워도 될것같슴다

CHAERYEON-LEE commented on src/components/common/Input/Input.tsx:
```diff
@@ -44,39 +41,41 @@ export const Input = ({
-      <p className={`${isValid ? styles.noError : styles.error}`}>
-        {validText}
-      </p>
+      {validText ? (
```
validText defaultProps로 주는것보다 이게 더 선언적인가욤?!!
저는 이렇게 쓴게 더 좋아보여욤!! 굳굳

pereng11 replied:
아 이거는 validtext를 주지 않은 경우에도 공간을 잡아먹어서 UI가 깨지는 바람에 수정했습니다!
defaultprops는 생각 못했는데 듣고보니 이렇게 하면 defaultprop대신 옵셔널 처리해도 될 것 같네요!

CHAERYEON-LEE commented on src/hooks/useInput.ts:
```diff
@@ -0,0 +1,28 @@
+import { useState } from 'react';
+
+const validateInput = (name: string, value: string) => {
+  return !inputValidationRegex[name]
```
이부분 혹시 
inputValidationRegex[name]
    ?  RegExp(inputValidationRegex[name]).test(value)
    :  true;
이렇게 쓰면 안되나욤
false 자리에 true가 적혀있어서 넘 이상한가

pereng11 replied:
아하 아니요 그게 더 좋은 것 같아요! 수정하겠습니다

CHAERYEON-LEE reviewed (commented):
넘모조아융

pereng11 committed:
refactor: Reflect feedback

- useInput hook의 유효성 검사 부분을 수정하였습니다
- 불필요한 스타일 코드를 삭제하였습니다
- 불필요한 defaultProps를 삭제하였습니다

pereng11 committed:
Merge branch 'Input-Jaeyeong' of github.com:pereng11/my-buddy-hero into Input-Jaeyeong

CHAERYEON-LEE reviewed (approved):
멋져욤
